test/infamy/netconf.py

```diff
@@ -107,11 +108,41 @@ def get_config(self, xpath):
     def get_config_dict(self, xpath):
         return self.get_config(xpath).print_dict()
 
-    def put_config(self, edit):
-        xml = "<config xmlns=\"urn:ietf:params:xml:ns:netconf:base:1.0\">" + edit + "</config>"
-        self.ncc.edit_config(xml)
+    def find_operation_dict(self, dictionary, keys=[]):
+        for key, value in dictionary.items():
+            if key != "@xmlns" and key != "operation":
+                keys.append(key)
+            if key == "operation":
+                return value, keys[-1] if keys else ""
+            elif isinstance(value, dict):
+                result = self.find_operation_dict(value, keys)
+                if result[0] != "merge":
+                    return result
+            elif isinstance(value, list):
+                for item in value:
+                    if isinstance(item, dict):
+                        result = self.find_operation_dict(item, keys)
+                        if result[0] != "merge":
+                            return result
+        return "merge", ""
+
+    def operation_config(self, xml, operation="merge", element=""):
+        if (operation != "merge" and element != ""):
+            pattern = fr'<{element}(?![^>]*nc:operation)(.*?)'
+            replacement = fr'<{element}\g<1> xmlns:nc="urn:ietf:params:xml:ns:netconf:base:1.0" nc:operation="{operation}"'
+            xml = re.sub(pattern, replacement, xml)
+        return xml
+
+    def put_config(self, edit, operation="merge", element=""):
+        edit = self.operation_config(edit, operation, element)
+        edit = "<config xmlns=\"urn:ietf:params:xml:ns:netconf:base:1.0\">" + edit + "</config>"
+        self.ncc.edit_config(edit)
 
     def put_config_dict(self, modname, edit):
+        # Extract operation and element from edit dictionary
+        operation, element = self.find_operation_dict(edit)
+
         mod = self.ly.get_module(modname)
         lyd = mod.parse_data_dict(edit, no_state=True)
-        return self.put_config(lyd.print_mem("xml", with_siblings=True, pretty=False))
+        xml = lyd.print_mem("xml", with_siblings=True, pretty=False)
+        self.put_config(xml, operation, element)
```

Comment:
I looked into this a bit, and it seems like `libyang` can do most of the heavy lifting for us here.

My suggestion would be to leave `put_config` and `put_config_dict` as-is. Then add a new method called `put_diff_dicts` which looks something like:

```python
def put_diff_dicts(self, modname, old, new):
    mod = self.ly.get_module(modname)
    oldd = mod.parse_data_dict(old, no_state=True)
    newd = mod.parse_data_dict(new, no_state=True)
    diff = oldd.diff(newd)
    xml = lyd.print_mem("xml", with_siblings=True, pretty=False)
    self.put_config(xml, operation, element)
```
This should create all of the needed YANG operation attributes etc.

In test cases, we can then use this by:
1. Getting the current config
2. Creating a deep copy of it
3. Remove the nodes and leafs we don't want
4. Call `put_diff_dicts` with the current running and the modified copy

What do you think?

Reply:
Alright, this seems like a bit simpler solution. But we'll discuss the implementation in person, I need to confirm what you think exactly.